plugins/com.google.cloud.tools.eclipse.appengine.deploy.ui/src/com/google/cloud/tools/eclipse/appengine/deploy/ui/DeployPreferencesDialog.java

```diff
@@ -0,0 +1,114 @@
+package com.google.cloud.tools.eclipse.appengine.deploy.ui;
+
+import org.eclipse.core.databinding.AggregateValidationStatus;
+import org.eclipse.core.databinding.DataBindingContext;
+import org.eclipse.core.databinding.observable.value.IValueChangeListener;
+import org.eclipse.core.databinding.observable.value.ValueChangeEvent;
+import org.eclipse.core.resources.IProject;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.jface.dialogs.Dialog;
+import org.eclipse.jface.dialogs.IDialogConstants;
+import org.eclipse.swt.SWT;
+import org.eclipse.swt.custom.CLabel;
+import org.eclipse.swt.custom.ScrolledComposite;
+import org.eclipse.swt.events.ControlEvent;
+import org.eclipse.swt.events.ControlListener;
+import org.eclipse.swt.graphics.Image;
+import org.eclipse.swt.graphics.Point;
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Control;
+import org.eclipse.swt.widgets.Shell;
+import org.eclipse.ui.ISharedImages;
+import org.eclipse.ui.PlatformUI;
+
+import com.google.common.base.Preconditions;
+
+public class DeployPreferencesDialog extends Dialog {
+
+  private static final Image ERROR_ICON = 
+      PlatformUI.getWorkbench().getSharedImages().getImageDescriptor(ISharedImages.IMG_OBJS_ERROR_TSK).createImage();
+
+  private DeployPreferencesPanel content;
+  private IProject project;
+
+  private CLabel validationMessage;
+
+  public DeployPreferencesDialog(Shell parentShell, IProject project) {
+    super(parentShell);
+    this.project = project;
+  }
+
+  @Override
+  protected void configureShell(Shell newShell) {
+    super.configureShell(newShell);
+    newShell.setText("Deployment preferences");
+  }
+
+  @Override
+  protected Control createDialogArea(final Composite parent) {
+    validationMessage = new CLabel(parent, SWT.LEFT);
+    GridData layoutData = new GridData(SWT.LEFT, SWT.CENTER, true, false);
+    layoutData.exclude = true;
+    validationMessage.setLayoutData(layoutData);
+    validationMessage.setVisible(false);
+
+    Composite dialogArea = (Composite) super.createDialogArea(parent);
+    ScrolledComposite scrolledComposite = new ScrolledComposite(dialogArea, SWT.V_SCROLL | SWT.H_SCROLL);
+    scrolledComposite.setLayoutData(new GridData(SWT.FILL, SWT.FILL, true, true));
+    scrolledComposite.setShowFocusedControl(true);
+    scrolledComposite.setExpandHorizontal(true);
+    scrolledComposite.setExpandVertical(true);
+    scrolledComposite.addControlListener(new ControlListener() {
+      @Override
+      public void controlResized(ControlEvent e) {
+        Point computeSize = DeployPreferencesDialog.this.getShell().computeSize(SWT.DEFAULT, SWT.DEFAULT);
```

Comment:
I don't think using the *Shell*'s size is correct.  You should be using the ScrolledComposite's content.